src/bin/migrate.rs

```diff
@@ -0,0 +1,257 @@
+//! This binary is used to migrate user IDs from base64 to hex encoding.
+use std::{path::Path, str::FromStr};
+
+use anyhow::{Context, Result};
+use famedly_sync::{
+	get_next_zitadel_user,
+	user::{ExternalIdEncoding, User as SyncUser},
+	zitadel::Zitadel as SyncZitadel,
+	Config,
+};
+use tracing::level_filters::LevelFilter;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+	// Config
+	let config_path =
+		std::env::var("FAMEDLY_SYNC_CONFIG").unwrap_or_else(|_| "./config.yaml".to_owned());
+	let config = Config::new(Path::new(&config_path))?;
+
+	// Tracing
+	let subscriber = tracing_subscriber::FmtSubscriber::builder()
+		.with_max_level(
+			config
+				.log_level
+				.as_ref()
+				.map_or(Ok(LevelFilter::INFO), |s| LevelFilter::from_str(s))?,
+		)
+		.finish();
+	tracing::subscriber::set_global_default(subscriber)
+		.context("Setting default tracing subscriber failed")?;
+
+	tracing::info!("Starting migration");
+	tracing::debug!("Old external IDs will be base64 decoded and re-encoded as hex");
+	tracing::debug!("Note: External IDs are stored in the nick_name field of the user's profile in Zitadel, often referred to as uid.");
+
+	// Zitadel
+	let mut zitadel = SyncZitadel::new(&config).await?;
+
+	// Detect external ID encoding based on a sample of users
+	let users_sample = zitadel.get_users_sample().await?;
+	let encoding = detect_database_encoding(users_sample);
```

Comment:
Why does the tool continue when the dominant encoding is `hex`? Right now it would not convert the possible remaining base64 user_ids to hex because it would expect hex in `create_user_with_converted_external_id`, or?

Reply:
The tool continues so that we can get logs about the particular cases, ie. we will see how many users don't match the db-detected encoding. Yes, it would not convert the possible `base64` user IDs to `hex` if `hex` was detected in DB heuristics. That's by design. Any better solution? Keep in mind that the result of a `hex` encoding is a subset of characters of the `base64` encoding. That's what prompted us with @tlater-famedly to implement DB heuristics.

Reply:
Yes makes sense. Just wondered why it would not stop because there is no need to migrate.